packetbeat: add support for NFS v3 and v4 protocols

kofemann opened the pull request:
This pull request is more an RFC. We will run it bit in production go get more experience. Comments are welcome. The example output:
```json
{
  "@timestamp": "2016-03-28T06:18:18.431Z",
  "beat": {
    "hostname": "localhost",
    "name": "localhost"
  },
  "count": 1,
  "dst": "127.0.0.1",
  "dst_port": 2049,
  "nfs": {
    "minor_version": 1,
    "opcode": "GETATTR",
    "status": "NFSERR_NOENT",
    "tag": "",
    "version": 4
  },
  "rpc": {
    "auth_flavor": "unix",
    "call_size": 200,
    "cred": {
      "gid": 500,
      "gids": [
        491,
        499,
        500
      ],
      "machinename": "localhost",
      "stamp": 4597002,
      "uid": 500
    },
    "reply_size": 96,
    "status": "success",
    "time": 25631000,
    "time_str": "25.631ms",
    "xid": "2cf0c876"
  },
  "src": "127.0.0.1",
  "src_port": 975,
  "type": "nfs"
}
```

elasticsearch-release commented:
Jenkins standing by to test this. If you aren't a maintainer, you can ignore this comment. Someone with commit access, please review this and clear it for Jenkins to run; then  say 'jenkins, test it'.

kofemann commented:
The one of the questions I would like to know: the other beats use field names with dots, but elasticsearch 2.0 does not allow that. So is it ok in beats?

kofemann commented:
Ok. Now it's in the shape, that  others can use it as well, IOW - this is a real pull request now.

kofemann commented:
Dashboard example
![screenshot from 2016-03-28 17-17-26](https://cloud.githubusercontent.com/assets/575965/14081134/60d7ab36-f509-11e5-83f0-b4f82da03473.png)

andrewkroh commented on packetbeat/protos/rpc/nfs.go:
```diff
@@ -0,0 +1,42 @@
+	xdr   Xdr
+	vers  uint32
+	proc  uint32
+	event *common.MapStr
```
I would recommend not using a pointer here. Maps are reference types so they are always passed by reference.

kofemann replied:
fixed.

andrewkroh commented on packetbeat/protos/rpc/nfs3.go:
```diff
@@ -0,0 +1,34 @@
+}
+
+func (nfs *Nfs) getV3Opcode() string {
+	if nfs.proc < 22 {
```
Should `22` be `len(nfs_opnum3)`?

kofemann replied:
fixed

andrewkroh commented on packetbeat/protos/rpc/nfs4.go:
```diff
@@ -0,0 +1,239 @@
+
+import "fmt"
+
+const OP_ACCESS = 3
```
Consider grouping these constants in a single `const` block. It's common to group related constants this way.

```
const (
    OP_ACCESS = 3
    OP_CLOSE = 4
)
```

I think you could also use [`iota`](https://golang.org/ref/spec#Iota) here.

```
const (
    OP_ACCESS = iota + 3
    OP_CLOSE
    OP_COMMIT
    ...
    OP_ILLEGAL = 10044
)
```

kofemann replied:
fixed

andrewkroh commented on packetbeat/protos/rpc/rpc.go:
```diff
@@ -0,0 +1,233 @@
+		}
+
+		if !is_last {
+			logp.Warn("rpc", "multifragment rpc message")
```
`logp.Warn` doesn't accept a debug selector argument from what I recall.

kofemann replied:
fixed

andrewkroh commented on packetbeat/protos/rpc/rpc.go:
```diff
@@ -0,0 +1,233 @@
+		// concatenate bytes
+		st.rawData = append(st.rawData, pkt.Payload...)
+		if len(st.rawData) > tcp.TCP_MAX_DATA_IN_STREAM {
+			logp.Debug("rpc", "Stream data too large, dropping TCP stream")
```
I would recommend creating a single `debugf` function for your package. And everywhere you write `logp.Debug("rpc", "some message")` replace it with `debugf("some message")`. It makes it easy to see what the debug selectors are for the package and it avoids the problem of typos with the selector.

Basically at the package level just add this:

`var debugf = logp.MakeDebug("rpc")`

kofemann replied:
fixed

andrewkroh commented on packetbeat/protos/rpc/rpc_message.go:
```diff
@@ -0,0 +1,117 @@
+
+var calls_seen = common.NewCache(1*time.Minute, 8192)
+
+func (msg *RpcMessage) fillEvent(event *common.MapStr, results publish.Transactions, size int) {
```
No need to use a MapStr pointer as mentioned earlier.

kofemann replied:
this was not clear to me. Thanks.

andrewkroh commented on packetbeat/protos/rpc/rpc_message.go:
```diff
@@ -0,0 +1,117 @@
+		rpc_prog_vers := msg.xdr.getUInt()
+		rpc_proc := msg.xdr.getUInt()
+
+		if rpc_prog == 100003 {
```
Seems like `100003` should be named constant?

kofemann replied:
fixed

andrewkroh commented:
Thanks for the contribution! Very cool screen shot. A couple miscellaneous points:

- Try running your package against `goimports`. It should change the ordering on some of the imports.
- Some of the variable names are using underscores. The Go convention is [mixedCase](https://golang.org/doc/effective_go.html#mixed-caps). Try running `golint` against your package.
- I see you added some pcaps. To ensure that there are no regressions we like to have some system tests in-place that replay the pcaps and check the outputs. Could you add some test cases like [these ones](https://github.com/elastic/beats/blob/master/packetbeat/tests/system/test_0032_dns.py) that use those pcaps?
- And finally, to make sure that everyone understands what the fields mean it would be great if you could document their meanings. We put each field into [fields.yml](https://github.com/elastic/beats/blob/master/packetbeat/etc/fields.yml) then we generate documentation from it by running `make update`.

kofemann commented:
I have addresses all issues you pointed out.. With style it was not clear, a if code used mixed style in various places, like  tcptuple.Dst_ip.String().  Any way, Tests added, rebased to current master,  branch updated.

andrewkroh commented:
> The one of the questions I would like to know: the other beats use field names with dots, but elasticsearch 2.0 does not allow that. So is it ok in beats?

Because field names containing dots are not allowed in ES 2.0 we do not use them in Beats. We use nested maps which gives the appearance of field names with dots because you can use dot-notation to Kibana and ES to reference the fields. For example:

```
event = common.MapStr{}
nfs = common.MapStr{}
event["nfs"] = nfs
nfs["version"] = 4
```

andrewkroh commented on packetbeat/tests/system/test_0061_nfs.py:
```diff
@@ -0,0 +1,58 @@
+
+        assert o["type"] == "nfs"
+        assert o["rpc.auth_flavor"] == "unix"
+	assert "rpc.time" in o
```
There appears to be a few tabs here that are throwing off the indentation. Please convert them to [spaces](http://legacy.python.org/dev/peps/pep-0008/#tabs-or-spaces). Thanks

kofemann replied:
sorry about that. fixed.

andrewkroh commented:
Looks good. I would like to get at least one more reviewer on this. @elastic/beats

kofemann commented:
I will create a separate PR for dashboard later on.

andrewkroh commented on packetbeat/etc/fields.yml:
```diff
@@ -1288,6 +1288,74 @@ trans_event:
           description: >
             The cursor identifier returned in the OP_REPLY. This must be the value that was returned from the database.
 
+    - name: rpc
```
Please run `make update` to re-generate the `packetbeat/docs/fields.asciidoc` document.

monicasarbu commented:
@kofemann Thank you for the complete PR and the nice dashboard. I have only a small concern. 

I think it would be easier for the users to understand if we rename the Packetbeat module to `nfs` to be the same with the name of the configuration section `nfs`. In this case, we can move all the RPC information from `rpc` under `nfs.rpc`.

kofemann commented:
@monicasarbu I had the same idea during prototyping was renaming it already several times :). The reason to call it rpc is that it will be easy to add other oncrpc/sunrpc based protocols as well. Nevertheless, I am fine to rename it. @andrewkroh any strong opinion?

andrewkroh commented:
I would rename to `nfs`. If in the future new RPC based protocols are added we can refactor a bit to reuse this code.

kofemann commented:
updated. The fields are still rpc.xxx and nfs.xxx. This makes more sense. You don't put **tcp** relates staff into **http** namspace

monicasarbu commented:
@kofemann Ok, I agree with you that it makes more sense to have `rpc` and `nfs` inside the nfs event type.

kofemann commented:
thanks! About dashboard: shold it go into packetbeat/etc/kibana ? Is there any conventions?

monicasarbu commented:
@kofemann I am working now at a simple way to import and export the Kibana dashboard. For now, you can only export all the Kibana dashboards and dependencies (visualizations, searches, index-patterns) using the python script from https://github.com/elastic/beats-dashboards/tree/master/save. Please send us a PR against the beats-dashboards repo with all the exported dashboards and dependencies. I can help you to sort them out once the PR is opened. Thanks!
